File: 21-01-2025_Linked_List_Group_Reverse.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return head
        
        current = head
        new_head = None
        tail = None

        while current:
            group_head = current
            prev = None
            next_node = None
            count = 0

            # Reverse k nodes
            while current and count < k:
                next_node = current.next
                current.next = prev
                prev = current
                current = next_node
                count += 1
            
            if new_head is None:
                new_head = prev
            
            if tail:
                tail.next = prev
            
            tail = group_head
        
        return new_head

File: 21-01-2025_Linked_List_Group_Reverse.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null) {
            return head;
        }

        ListNode current = head;
        ListNode newHead = null;
        ListNode tail = null;

        while (current != null) {
            ListNode groupHead = current;
            ListNode prev = null;
            ListNode nextNode = null;
            int count = 0;

            // Reverse k nodes
            while (current != null && count < k) {
                nextNode = current.next;
                current.next = prev;
                prev = current;
                current = nextNode;
                count++;
            }

            if (newHead == null) {
                newHead = prev;
            }

            if (tail != null) {
                tail.next = prev;
            }

            tail = groupHead;
        }

        return newHead;
    }
}
